File: .migration-backup/artifacts/api-server/src/routes/analytics.ts
import { Router, type IRouter } from "express";
import { and, count, desc, eq, sql } from "drizzle-orm";
import { db, conversations, messages, handoffs } from "../lib/db";
import { requireJwt } from "../middlewares/auth";
import { timeAgo } from "../lib/serialize";

const router: IRouter = Router();

export async function buildAnalytics(tenantId: number) {
  const [msgCountRow] = await db
    .select({ n: count() })
    .from(messages)
    .innerJoin(conversations, eq(messages.conversationId, conversations.id))
    .where(and(eq(conversations.tenantId, tenantId), eq(messages.role, "user")));
  const totalQuestions = Number(msgCountRow?.n ?? 0);

  const [handoffCountRow] = await db
    .select({ n: count() })
    .from(handoffs)
    .where(eq(handoffs.tenantId, tenantId));
  const totalHandoffs = Number(handoffCountRow?.n ?? 0);

  const deflectionRate =
    totalQuestions > 0
      ? Math.max(0, Math.min(100, ((totalQuestions - totalHandoffs) / totalQuestions) * 100))
      : 0;

  // Channel distribution from conversations.
  const channelRows = await db
    .select({ channel: conversations.channel, n: count() })
    .from(conversations)
    .where(eq(conversations.tenantId, tenantId))
    .groupBy(conversations.channel);

  // Top user-questions by frequency.
  const topRows = await db.execute(sql`
    SELECT m.content AS name, COUNT(*) AS count
    FROM messages m
    JOIN conversations c ON c.id = m.conversation_id
    WHERE c.tenant_id = ${tenantId} AND m.role = 'user'
    GROUP BY m.content
    ORDER BY count DESC
    LIMIT 5
  `);
  const topQuestions = (topRows as unknown as { rows: { name: string; count: number }[] }).rows.map(
    (r) => ({ name: r.name, count: Number(r.count) }),
  );

  // Unanswered: pending handoffs with last user message.
  const pending = await db
    .select()
    .from(handoffs)
    .where(and(eq(handoffs.tenantId, tenantId), eq(handoffs.status, "pending")))
    .orderBy(desc(handoffs.createdAt))
    .limit(10);
  const unanswered = pending.map((h) => ({
    id: h.id,
    question: h.question,
    channel: h.channel,
    language: h.language,
    confidence: 0,
    timeAgo: timeAgo(h.createdAt),
  }));

  // Average CSAT score across resolved handoffs.
  const [csatRow] = await db
    .select({ avg: sql<number>`COALESCE(AVG(${handoffs.csatScore}), 0)` })
    .from(handoffs)
    .where(and(eq(handoffs.tenantId, tenantId), sql`${handoffs.csatScore} IS NOT NULL`));
  const feedbackScore = Number(csatRow?.avg ?? 0);

  return {
    kpis: {
      total_questions: totalQuestions,
      deflection_rate: Number(deflectionRate.toFixed(1)),
      cost_savings: Math.round(totalQuestions * 0.5 * 100) / 100,
      feedback_score: Number(feedbackScore.toFixed(2)),
    },
    top_questions: topQuestions,
    channel_distribution: channelRows.map((r) => ({ name: r.channel, value: Number(r.n) })),
    unanswered,
  };
}

router.get("/analytics", requireJwt, async (req, res) => {
  return res.json(await buildAnalytics(req.auth!.tid));
});

router.get("/analytics/dashboard", requireJwt, async (req, res) => {
  return res.json(await buildAnalytics(req.auth!.tid));
});

export default router;

File: artifacts/api-server/app/services/analytics.py
from datetime import datetime, timezone
from sqlalchemy import func, select, text
from sqlalchemy.ext.asyncio import AsyncSession
from app.models import Conversation, Handoff, Message
from app.schemas import AnalyticsResponse, KPIs, TopQuestion, ChannelDist, UnansweredItem
from app.services.serialize import time_ago


async def build_analytics(db: AsyncSession, tenant_id: int) -> AnalyticsResponse:
    # Total user messages (questions)
    conv_ids = select(Conversation.id).where(Conversation.tenant_id == tenant_id)
    total_q_result = await db.execute(
        select(func.count(Message.id)).where(
            Message.conversation_id.in_(conv_ids),
            Message.role == "user",
        )
    )
    total_questions = total_q_result.scalar_one() or 0

    # Deflection rate = (questions not escalated) / total
    pending_result = await db.execute(
        select(func.count(Handoff.id)).where(Handoff.tenant_id == tenant_id)
    )
    total_handoffs = pending_result.scalar_one() or 0
    deflection_rate = round((1 - total_handoffs / total_questions) * 100, 1) if total_questions else 0.0

    cost_savings = round(total_questions * 0.5, 2)

    # Feedback score
    score_result = await db.execute(
        select(func.avg(Handoff.csat_score)).where(
            Handoff.tenant_id == tenant_id,
            Handoff.csat_score.is_not(None),
        )
    )
    avg_score = score_result.scalar_one()
    feedback_score = round(float(avg_score), 2) if avg_score else 0.0

    # Top questions (user messages)
    top_q_result = await db.execute(
        select(Message.content, func.count(Message.id).label("cnt"))
        .where(Message.conversation_id.in_(conv_ids), Message.role == "user")
        .group_by(Message.content)
        .order_by(text("cnt DESC"))
        .limit(5)
    )
    top_questions = [TopQuestion(name=r[0], count=r[1]) for r in top_q_result.all()]

    # Channel distribution
    ch_result = await db.execute(
        select(Conversation.channel, func.count(Conversation.id).label("cnt"))
        .where(Conversation.tenant_id == tenant_id)
        .group_by(Conversation.channel)
    )
    channel_distribution = [ChannelDist(name=r[0], value=r[1]) for r in ch_result.all()]

    # Unanswered (pending handoffs with low-confidence reason)
    unanswered_result = await db.execute(
        select(Handoff)
        .where(Handoff.tenant_id == tenant_id, Handoff.status == "pending")
        .order_by(Handoff.created_at.desc())
        .limit(10)
    )
    unanswered = []
    for h in unanswered_result.scalars().all():
        unanswered.append(UnansweredItem(
            id=h.id,
            question=h.question or "",
            channel=h.channel,
            language=h.language,
            confidence=0.0,
            time_ago=time_ago(h.created_at),
        ))

    return AnalyticsResponse(
        kpis=KPIs(
            total_questions=total_questions,
            deflection_rate=deflection_rate,
            cost_savings=cost_savings,
            feedback_score=feedback_score,
        ),
        top_questions=top_questions,
        channel_distribution=channel_distribution,
        unanswered=unanswered,
    )
